lib/Sema/CSGen.cpp

```diff
@@ -2827,7 +2827,8 @@ namespace {
 
     Type visitDiscardAssignmentExpr(DiscardAssignmentExpr *expr) {
       auto locator = CS.getConstraintLocator(expr);
-      auto typeVar = CS.createTypeVariable(locator, TVO_CanBindToNoEscape);
+      auto typeVar = CS.createTypeVariable(locator, TVO_CanBindToNoEscape |
+                                                    TVO_CanBindToHole);
```

Comment:
Is it always safe to assume that we'll get a fix every time this is type variable becomes a hole? I'm asking because otherwise we'll produce and apply solutions with UnresolvedType which in production builds would crash somewhere in SILGen...

Reply:
I think the thing we can rely on is the `discard_expr_outside_of_assignment` diagnostic. `MiscDiagnostics` reports this error error message for any `_` that is not the destination of an assignment, regardless of the inferred type, so introducing a hole won't change that behavior. When encountering an `AssignExpr`, it will look for any `_` in the destination and mark them as "acceptable".
```cpp
      // If we have an assignment expression, scout ahead for acceptable _'s.
      if (auto *AE = dyn_cast<AssignExpr>(E))
        markAcceptableDiscardExprs(AE->getDest());

      /// Diagnose a '_' that isn't on the immediate LHS of an assignment.
      if (auto *DAE = dyn_cast<DiscardAssignmentExpr>(E)) {
        if (!CorrectDiscardAssignmentExprs.count(DAE) &&
            !DAE->getType()->hasError())
          Ctx.Diags.diagnose(DAE->getLoc(),
                             diag::discard_expr_outside_of_assignment);
      }
```

The only way we'd get away with inferring a hole for `_` and not reporting an error message is if the `_` _is_ on the left side of an assignment and we inferred a hole based on the `TVO_CanBindToHole` flag from CSGen. I can't think of a way this could happen, because if we infer a hole for the underscore in `_ = <some expression>`, it means that we couldn't infer the type of `<some expression>`, and we'd report an error message for that. Can you think of anything I'm missing in this assumption? 

I do think this assumption is kind of subtle and could be more explicit. One thing we could do instead is the same thing we do for generic parameters (record a fix in `TypeVariableBinding::attempt` when we use the hole default), and then in the diagnostic figure out whether to report `diag::pattern_in_expr` or `diag::discard_expr_outside_of_assignment`. I think we could also report `diag::pattern_in_expr` in CSGen when we encounter an `UnresolvedPatternExpr` to get that case out of the way. What do you think?

Reply:
I think it's a reasonable assumption that if right-hand side of the assignment is a hole we would have produced a fix already for that. 

To be on the safe side we could record an error just like we do for generic parameters in `TypeVariableBinding::attempt` and skip the diagnostic if there are other fixes available. I'll leave this up to you to decide.

Reply:
I started to implement the safe thing thinking that I could remove the diagnostic from `MiscDiagnostics`, but then I realized that that wouldn't handle invalid underscores that we do infer a type for. Going to leave it as is for now

Reply:
Sounds good!